### DISCUSSION TITLE
gnpy-transmission-example customisation query

### DISCUSSION THREAD

**Discussion by uceefsi** [Telecominfraproject/oopt-gnpy#416] [Q&A] [answered]
Hi team

I hope this message finds you well.

First of all, allow me to thank you for this great and useful tool.  
I'm currently learning and evaluating the tool in order to simulate an optical path composed of Transceivers, ROADMS, Optical Amplifiers, and fibre across several nodes.

I've been learning using the Docker installation and running "gnpy-transmission-example", and I must say i'm very impressed and helpful with the tool. A lot of potential indeed. Fantastic work you all :) 

I'm currently trying to map some additional Optical Amplifiers and adding some more data to create some practical simulations but i'm getting an error when running the script. A couple of questions if I may please:

1.  I'm getting "inf" for the OSNR and SNR of the first Transceiver :(

     Propagating with input power = 0.00 dBm: Transceiver trx Equinix_LD8_DC OSNR ASE (0.1nm, dB): inf
     OSNR ASE (signal bw, dB): inf
     SNR total (signal bw, dB): inf

2.  I'm using as input Excel file "Inf_Topology_V003.xls" (attached) and I've allocated in "amp_type" fields all to be "operator_model_example", but i see the script sometimes uses it, and sometimes it doesn't. Is there any way to force it to use it?

Below just an extract of the output, but happy to share all the content. Once again, thank so much and what a great project, good work team !!!

Any comments/suggestions must welcome:

Best Regards,

Fernando Silva 
uceefsi@ucl.ac.uk

[Inf_Topology_V003.xls](https://github.com/Telecominfraproject/oopt-gnpy/files/6851163/Inf_Topology_V003.xls)
[gnpy-transmission-example_output.txt](https://github.com/Telecominfraproject/oopt-gnpy/files/6851165/gnpy-transmission-example_output.txt)

Extract of the output:

# gnpy-transmission-example -e eqpt_config.json Inf_Topology_V003.xls "Equinix_LD8_DC" "Telehouse_North_DC"
There are 76 channels propagating
Power mode is set to True
=> it can be modified in eqpt_config.json - Span

There are 5 fiber spans over 156 km between trx Equinix_LD8_DC and trx Telehouse_North_DC

Now propagating between trx Equinix_LD8_DC and trx Telehouse_North_DC:

Propagating with input power = 0.00 dBm:
Transceiver trx Equinix_LD8_DC
  OSNR ASE (0.1nm, dB):      inf                                         "<-- expecting anything but inf (Question1)"
  OSNR ASE (signal bw, dB):  inf                                        "<-- expecting anything but inf (Question 1)"
  SNR total (signal bw, dB): inf                                          "<-- expecting anything but inf (Question 1)"
  SNR total (0.1nm, dB):     inf                                           "<-- expecting anything but inf (Question 1)"
  CD (ps/nm):                0.00
  PMD (ps):                  0.00
Roadm roadm Equinix_LD8_DC
  effective loss (dB):  20.00
  pch out (dBm):        -20
Edfa east edfa in Equinix_LD8_DC to BT_Vauxhall
  type_variety:           operator_model_example
  effective gain(dB):     20.69
  (before att_in and before output VOA)
  noise figure (dB):      5.60
  (including att_in)
  pad att_in (dB):        0.00
  Power In (dBm):         -1.19
  Power Out (dBm):        19.51
  Delta_P (dB):           1.0
  target pch (dBm):       1.0
  effective pch (dBm):    0.69
  output VOA (dB):        0.00
Fiber          fiber (Equinix_LD8_DC → BT_Vauxhall)-NC-DW-6907
  type_variety:                SSMF
  length (km):                 70.00
  pad att_in (dB):             0.00
  total loss (dB):             13.97
  (includes conn loss (dB) in: 0.00 out: 0.00)
  (conn loss out includes EOL margin defined in eqpt_config.json)
  pch out (dBm): -13.27
Edfa Edfa0_fiber (Equinix_LD8_DC → BT_Vauxhall)-NC-DW-6907
  type_variety:           std_low_gain                                                    "<-- Expecting "operator_model_example" (Question 2)"
  effective gain(dB):     13.27
  (before att_in and before output VOA)
...

**Comment by ojnas** [comment]
Hi @uceefsi,

thanks for the nice feedback and for providing a detailed error description. Regarding the first issue (getting "inf" OSNR), I just tried it with a Docker installation and I can reproduce the error.

@jktjkt, it looks like the script in the Docker container uses an old version of the code that was installed by pip:

```
root:/shared/example-data# pip show gnpy
Name: gnpy
Version: 2.2.0.0a1.dev6
```

**Reply by uceefsi** [reply]
Hej Jonas, 
Thanks so much for your prompt reply, It's fantastic to see the tool it's in fact maintained and in continuous development.  
@jktjkt, appreciate if you let me/us know once the new version has been uploaded so I can pull the new image.

In regards to the second question I've raised, I think I found the answer and it has to do with the auto-design option with the power mode; if I understand correctly, on auto mode the amplifier gain would equal the span loss so the amp will be selected accordingly.  I also learnt I can set the "Booster" and "Preamp" restriction to the amp I want in the design (in the xls Nodes tab), and that will force it to be used in the design... so happy days !

I'll continue conducting tests and I'll report back in case I ran into trouble again ;)

Once again, great tool and great collaboration team

Best Regards, 

Fernando Silva
uceefsi@ucl.ac.uk

**Reply by jktjkt** [reply]
Docker pulls the tag called `latest` by default, and that one is currently not being updated (I'll fix that; I've now at least removed that tag so as not to offer an outdated version by default). As a quick workaround, when you `docker run`, try using an explicit version of a Docker image, e.g., `telecominfraproject/oopt-gnpy:v2.3.1`. That should do the trick.

**Comment by ojnas** [comment, marked as answer]
@uceefsi, regarding the second issue, you are correct that auto-design has selected those std_low_gain amps. But auto-design should (and does) not change the type of amps specified for a node by the user in the Eqpt sheet of the topology excel file. The problem here is that each row in the Eqpt sheet only specifies the amps (egress/ingress) in one end (Node A) of the link. To fully specify all amps in the topology you would need an additional row for each link in the opposite direction (I have attached a modified excel file where this is done).

It is also true that you can use Booster/Preamp restriction in the Nodes sheet to force the selection of amps. Note, however, that when you have actually specified the amp in the Eqpt sheet, this will override the restriction and you are not really meant to mix these methods. Also, specifying restrictions only works for ROADM nodes and not for ILA nodes.

[Inf_Topology_V003_modified.xls](https://github.com/Telecominfraproject/oopt-gnpy/files/6866822/Inf_Topology_V003_modified.xls)

**Comment by uceefsi** [comment]
Hej @ojnas  

I hope this message finds you well. 
Thanks so much for your answer again, that certainly makes more sense ;).  It's starting to look more to what I was expecting in the design. Fantastic help and guidance again, much appreciated !

Going through the new results, I'm curious about another behaviour, but this time between the launch power and the power per channel.  

When I'm executing the script, I'm setting launch power to 1.0dBm; how I understand it, this is (or should be) the launch power of the originating Transceiver, correct?  If that's correct, what I find a bit odd is that I want to target an output channel power to 3.5dB ("per channel"), and I have set this up in the "eqpt_config.json"-roadm section.  (attached slide highlighting) detail.  

What I'm getting in the script results however, is that somehow whatever launch power is defined by the "-po" argument (0dBm by default) is being overwritten/used as the "target pch" / "effective pch" across all the Amp calculations, when in reality the output power channel I'm trying to aim in the design is 3.5dB "per channel".   Attached a ppt slide highlighting the detail, as well as the eqpt_config.json and xls topology used for completion.

Greatly appreciate if you can please point me in the right direction again and once again many thanks for all the support (and apologies for raising so many questions, but I'm just loving the tool ;)

All the best over there and best regards,

Fernando Silva
uceefsi@ucl.ac.uk 

[GNPy Launch power vs channel power.pptx](https://github.com/Telecominfraproject/oopt-gnpy/files/6868327/GNPy.Launch.power.vs.channel.power.pptx)
[gnpy-transmission-example output with lauch power set to 1dB.txt](https://github.com/Telecominfraproject/oopt-gnpy/files/6868328/gnpy-transmission-example.output.with.lauch.power.set.to.1dB.txt)
[Inf_Topology_V016.xls](https://github.com/Telecominfraproject/oopt-gnpy/files/6868332/Inf_Topology_V016.xls)

**Reply by ojnas** [reply]
Hi @uceefsi,

what you're seeing is the expected behavior. The `target_pch_out_db` parameter in the Roadm section just defines the output power of the WSS and into the booster. The power per channel after the booster and into the fiber is defined by the `power_dbm` parameter in the `SI` section of the eqpt file (which can be overridden by the `-po` argument) and the `delta_p` of the amp in the topology file.

**Reply by uceefsi** [reply]
Hej @ojnas 
Thanks again for your valuable comments and clarifications.  It makes more sense now ;)
It's a shame with the docker install i can't really see or follow step by step "gnpy-transmission-example" script since i can't find it in the docker folders ;(
I'll continue working with the installed version, and trying to understand the parameters to accommodate them in a design and look at further use cases.  
Again, much appreciated, fantastic work and collaboration !

**Reply by ojnas** [reply]
Hej,

@jktjkt, it seems like the latest release version was not tagged as "latest" when pushed to docker hub so `docker pull telecominfraproject/oopt-gnpy` pulls an old version.

@uceefsi, sorry for the late reply. If you want to run a docker container with the most up to date master version of the repo, you can use `telecominfraproject/oopt-gnpy:master`. And the `gnpy-transmission-example` script is just a wrapper around the `transmission_main_example` function in `gnpy/tools/cli_examples.py`.

**Reply by jktjkt** [reply]
@ojnas , it looks like I commented without seeing your response from today. I've removed the `latest` tag, and I'll change the CI configuration so that all pushes to tags update the `latest` tag. (We will have issues with this if we ever start supporting multiple releases I guess...).

@uceefsi , in the Docker container the GNPy sources are available below `/usr/local/lib/python3.9/site-packages/gnpy/`. However, keep in mind that the Docker environment is rather restrictive. If you would like to dive into GNPy's sources, perhaps a git checkout and an installation into a Python `venv` will work better.

**Reply by ojnas** [reply]
@jktjkt, do we want all pushes to update the "latest" tag? Maybe it would be more logical to only tag the latest release version (currently 2.3.1) with "latest". Otherwise, `docker pull telecominfraproject/oopt-gnpy` and `pip install gnpy` would install different versions, which may be confusing.

**Reply by jktjkt** [reply]
Yup, I'm limiting that to "releases only" (`^refs/tags/v.*$`) -- see https://github.com/Telecominfraproject/oopt-gnpy/pull/417

**Reply by uceefsi** [reply]
Hej @ojnas  / @jktjkt 

I hope you're all keeping well.  Thanks again for your valuable updates and keep the application up to date. 

I confirm if I run Docker image v2.3.1 I don't see the "inf" values in the xSNR, so that's great.  I do see however a couple of things:

1. No value shown for Chromatic Dispersion "CD" nor Polarisation Mode Dispersion "PMD"  for the first hop 
2. I see now a "G" before "SNR".  (The results compared to other versions is exactly the same but the label/naming has changed)

I attach the detail of questions 1 and 2, in "GNPy Queries docker version 2.3.1.png" and the excel input file "Inf_Topology_V022.xls"

![GNPy Queries Docker version 2 3 1](https://user-images.githubusercontent.com/86266872/136848257-1428b4e8-4664-4d73-ba7d-574d582a3cf8.PNG)
I attach for reference purposes the input excel file i'm using for testing "Inf_Topology_V022.xls
[Inf_Topology_V022.xls](https://github.com/Telecominfraproject/oopt-gnpy/files/7325034/Inf_Topology_V022.xls)
x"

As always, your support and assistance is much appreciated.  Keep safe and many thanks again

Best Regards, 

Fernando Silva P
uceefsi@ucl.ac.uk

**Reply by ojnas** [reply]
Hej @uceefsi,

1. The CD and PMD are zero at the source trx simply because the signal has not propagated through any element with CD or PMD at that point. I guess this output is kind of redundant and could be removed to avoid confusion.
2. The "G" stands for "Generalized". We changed the output to say GSNR since that has become a well established term.

**Reply by ojnas** [reply]
@uceefsi, meanwhile there's been a new release (v2.4) and the "latest" tag is now correctly added to the docker image so you can just do `docker pull telecominfraproject/oopt-gnpy` again to get the latest version.

**Reply by jktjkt** [reply]
@uceefsi: And just to add to @ojnas' answer above, GSNR is explained in e.g. the [Rec. ITU-T G.977.1 (10/2020)](https://www.itu.int/rec/dologin_pub.asp?lang=e&id=T-REC-G.977.1-202010-I!!PDF-E&type=items). A short summary is that GNPy uses the term **OSNR** to refer to the OSNR from amplifiers' ASE, whereas the **GSNR** also takes into account the NLI.

**Reply by uceefsi** [reply]
Hej @ojnas / @jktjkt 
Thanks so much again for your valuable comments.  I'm a happy man now ;)
I confirm I've tested the docker pull as instructed and now that version also comes with the fix.  Much appreciate it again.  I'll continue using it and validating my internal tests, much appreciated again.
Good evening over there
